### ISSUE TITLE
Add token to a GraphQL/REST/gRPC/Java query to track queries for a single unit of work

### ISSUE THREAD

**Issue by novoj** [FgForrest/evitaDB#148]
We want clients to provide a new grouping information - a random token that could be used to group multiple requests with a single unit of work - such as rendering a single page. This unit of work could be tracked in monitoring and used to "debug" the frontstore and evaluate whether the client is using the API correctly.

For example, we might discover that a single page is rendering dozens of overlapping queries instead of one or two. We could also track the impact of prefetching and other techniques common in the modern JS world.

The database could also track the queries for each unit of work and create e-tags that allow frontstores to securely cache data locally, radically reducing latency and increasing throughput - static but up-to-date e-shops, yay!

cc @lukashornych

**Comment by novoj** [comment]
As discussed in all protocols there is a way how to propagate metadata from the clients. We need two tokens instead of one:

- `CLIENT_ID`: string that will be constant per a connected client, example values: `Next.JS Middleware`, `evitaDB console` etc.
- `REQUEST_ID`: a randomized token - preferably UUID that will connect all queries and mutations issued by the client in a single unit of work, that is controlled by the client (not the server); there might be different request ids within single evita session and also there might be same request id among multiple different evita sessions

We just need to acquire those tokens - the issue doesn't involve any actual operations with those data. But in the future there might be features connected with them:

- **performance analysis**: we could be able to help frontend developers to identify performance problems if we could see how many and which queries are issued for rendering single (or particular) page
- **query visualization**: we could provide a facility in planned evitaDB console that would trap and visualize the request processing connected with particular client and aggregate data to meaningful request packs

We need to be able to gather the data as soon as possible in order to integrate id issuing logic in currently developed client applications.

**Comment by novoj** [comment]
Initial proposal is in the class: https://github.com/FgForrest/evitaDB/blob/c86bc95df9c9425c5b73e9f39d0408b96b321806/evita_api/src/main/java/io/evitadb/api/ClientContext.java#L53

The expected usage is demonstrated by test: https://github.com/FgForrest/evitaDB/blob/c86bc95df9c9425c5b73e9f39d0408b96b321806/evita_functional_tests/src/test/java/io/evitadb/api/ClientContextTest.java#L73-L74

Can you @lukashornych @Khertys try to implement those in external APIs?

I've already added the basic integration to the `EvitaClient`. The clientId is part of the `EvitaClientConfiguration` and the client logic should manually set only the `requestId` as it suits him/her.

**Comment by lukashornych** [comment]
I've studied the prototype and I have a few questions (maybe I'm missing something):
- shouldn't this be stored in evitaDB engine globally, so that it can be viewed even with other clients, e.g. the planned evitaLab? For example to preview calls from FE application in the evitaLab even though it is different client? How would you access the data to preview otherwise?
- if EvitaClient stores this locally, why we need gRPC to support this (if we omit that someone can use the gRPC directly)?
- what if client using the EvitaClient or GQL library wants to run some part of inner evitaDB calls in separate threads using executor? Wouldn't these other threads lose access to the ids? Because I guess that the executor threads wouldn't be child threads of the main called one. This could maybe be partially solved by the ScopedValues, but still it would need to be explicitly handled. Even when we would want to read these ids (and data with them) to debug them via same client, different calls could receive different separate ids, wouldn't it?

I understand that you don't want to make the evitaDB API ugly by need to pass these ids everywhere, I just don't see how this local implementation could work :thinking: 
I will think more about different approach of storing, but currently I can't come up with something useful.

**Comment by novoj** [comment]
The last comment was discussed on Discord. The summary is here:

- I'll try to rewrite `ClientContext` to be part of `EvitaContract` and `EvitaSessionContract`.
- EvitaClient has only a partial implementation - I need @Khertys to propagate the identifiers to the server over the gRPC protocol and open the same ClientContext on the server side.
- The last comment is more complex to explain:
   - Currently all mutations/queries are executed single-threaded on the evitaDB side - if it gets parallelized it will be with JDK 20 scoped values which work correctly with virtual threads and the parallelization will take advantage of virtual threads.
   - evitaLab will not use the current API to get the "records" of the queries - the current API is designed for capturing the data.
   - There will be a separate API that will be able to browse through the records of queries/mutations, which will contain the captured client/request ID metadata and allow filtering on it.

All issues were resolved in the discussion.

**Comment by lukashornych** [comment]
Yes, that's what i was talking about.

**Comment by lukashornych** [comment]
@novoj Support in GQL and REST APIs is done.

The GQL API expects following extension object:
```json
"clientContext": {
  "clientId": "Next.JS Middleware",
  "requestId": "05e620b2-5b40-4932-b585-bf3bb6bde4b3"
}
```
Both `clientId` and `requestId` are optional.
The defaults are: 
```json
"clientContext": {
  "clientId": "unknownGraphQLClient",
  "requestId": null
}
```

The REST API expects following HTTP headers:
```
X-EvitaDB-ClientID: Next.JS Middleware
X-EvitaDB-RequestID: 05e620b2-5b40-4932-b585-bf3bb6bde4b3
```
Both headers are optional.
The defaults are:
```
X-EvitaDB-ClientID: unknownRestClient
X-EvitaDB-RequestID: null
```

**Comment by novoj** [comment]
@lukashornych Can you please set up a new chapter in `documentation/user/en/operate/monitor.md` and describe it there withi `LanguageSpecific` secion?! I will add a `Java` description here eventually.

**Comment by lukashornych** [comment]
@novoj Documentation is done. I left the Java version up to you as you wanted..

**Comment by lukashornych** [comment]
Some points from discord discussion:
- we agreed  on more complex client ID format: `{protocol}|{client address}|{client-defined id}`
  - the `{protocol}` will be static identification of a API: `GQL`, `GRPC` or `REST`
  - the `{client address}` will be incoming IP address of the input request
  - the `{client-defined id}` will be something the user will specify to identify its concrete application, e.g., `my-shop.com`
    - there will only basic ASCII characters supported for security as we will want to safely display it in the lab, also it will simply the names hopelly
  - workflow is as follows:
    - in EvitaClient or GQL client, the client will specify only his identification like `my-shop.com`
    - each API will then convert it into the above mentioned format and includes the user-defined client ID at the end
    - if an user uses the evitaDB in embedded version, it is up to him to use format whichever he/she wants, we don't have much control over it other than validation of allowed characters
  - this will helps as to group clientIds by protocol and address and so on in lab in the future

Next week, I'll prepare tooling for it so we can fully implement id.

**Comment by novoj** [comment]
The embedded version doesn't impose a security threat. The sanitization it this particular case doesn't make sense.

**Comment by lukashornych** [comment]
Well yeah, as I'm thinking about it now, the implementation that I had in mind wouldn't had place for it anyway, so it all good. Thanks for the point @novoj.

**Comment by lukashornych** [comment]
Hmm, but it kinda does, because even with embedded evitaDB you can still expose API for lab. Or did you wanted the sanitization for something else than the lab?

**Comment by novoj** [comment]
Yes, but the possible attack scenario is that malicious client injects something potentially dangerous to the clientId string. If it does the application that embeds the evitaDB, it's not a client - it's a server and the author shoots himself in the foot. When embedded evitaDB opens its ports, the requests arrived through these ports originate in real clients (maybe dangerous), but this way is taken care of and sanitized. Am I wrong?

**Comment by lukashornych** [comment]
Yeah, you're right. Of course. Just thinking that if we have the sanitizer already implemented and sort of built-in in the `ClientContext` why not just validate it inside the `Context` class every time instead of relying only on APIs to sanitize it. It will not represent much overhead for us, I think.

**Comment by novoj** [comment]
We might, but we only need to sanitize the "client" part - the IP / protocol part of string doesn't need to be sanitized. I still think that securing on the boundary is more correct.

**Comment by lukashornych** [comment]
@novoj Everything we talked about has been implemented. The GraphQL and REST APIs are documented in user docs. There is a TODO for you to write the Java user docs. 

One question though, in the issue there is task `document GRPC integration`... does that mean in user docs or Java docs? Because we don't have gRPC lang in user docs.

**Comment by novoj** [comment]
I propose to document it here: https://evitadb.io/documentation/use/connectors/grpc

**Comment by lukashornych** [comment]
cc @Khertys
